src/runtime/policy/policy.ts

```diff
@@ -0,0 +1,270 @@
+ * actual functionality other than validating the input is correct.
+ */
+export class Policy {
+  constructor(
+      readonly name: string,
+      readonly targets: PolicyTarget[],
+      readonly configs: PolicyConfig[],
+      private readonly annotationRefs: AnnotationRef[]) {
+    this.validateEgressType();
+  }
+
+  get description(): string | null {
+    const description = this.annotationRefs.find(a => a.name === 'intendedPurpose');
+    return description ? description.params['description'] as string : null;
+  }
+
+  get egressType(): PolicyEgressType {
+    const egressType = this.annotationRefs.find(a => a.name === 'egressType');
+    return PolicyEgressType[egressType.params['type'] as string];
+  }
+
+  get annotations() {
+    const specialAnnotations = ['intendedPurpose', 'egressType'];
+    return this.annotationRefs.filter(annotation => !specialAnnotations.includes(annotation.name));
+  }
+
+  private validateEgressType() {
+    const egressType = this.annotationRefs.find(annotation => annotation.name === 'egressType');
+    if (egressType) {
+      checkValueInEnum(egressType.params['type'], PolicyEgressType);
+    }
+  }
+
+  toManifestString(): string {
+    return [
+      ...this.annotationRefs.map(annotation => annotation.toString()),
+      `policy ${this.name} {`,
+      ...flatMap(this.targets, target => indentLines(target.toManifestString())),
+      ...flatMap(this.configs, config => indentLines(config.toManifestString())),
+      '}',
+    ].join('\n');
+  }
+
+  static fromAstNode(
+      node: AstNode.Policy,
+      buildAnnotationRefs: (ref: AstNode.AnnotationRef[]) => AnnotationRef[]): Policy {
+    const annotationRefs = buildAnnotationRefs(node.annotationRefs);
+
+    checkNamesAreUnique(node.targets.map(target => ({name: target.schemaName})));
+    const targets = node.targets.map(target => PolicyTarget.fromAstNode(target, buildAnnotationRefs));
+
+    checkNamesAreUnique(node.configs);
+    const configs = node.configs.map(config => PolicyConfig.fromAstNode(config));
+
+    return new Policy(node.name, targets, configs, annotationRefs);
+  }
+}
+
+class PolicyTarget {
+  constructor(
+      readonly schemaName: string,
+      readonly fields: PolicyField[],
+      private readonly annotationRefs: AnnotationRef[]) {
+    this.validateRetentions();
+    // TODO(b/157605585): Validate maxAge annotation.
+  }
+
+  get retentions() {
+    return this.annotationRefs
+        .filter(annotation => annotation.name === 'allowedRetention')
+        .map(annotation => ({
+          medium: PolicyRetentionMedium[annotation.params['medium'] as string],
+          encryptionRequired: annotation.params['encryption'] as boolean,
+        }));
+  }
+
+  get annotations() {
+    return this.annotationRefs.filter(annotation => annotation.name !== 'allowedRetention');
+  }
+
+  private validateRetentions() {
+    // Check retention annotations are valid.
```

Comment:
nit - this comment is redundant. maybe worth explaining what is an invalid retention, or remove it.
possible annotations enhancement - if we allow multiple of the same annotation, maybe also configure which param must be unique?

Reply:
Done.

I'm not sure it's worth it. Even in these couple of examples where I check it manually, there's some extra complexity (e.g. completely duplicate `@allowedUsage` annotations aren't allowed, but different ones with the same label are, provided that the usage type is different). I'm not sure how much use there'd be.